python-checks/src/main/java/org/sonar/python/checks/FStringNestingLevelCheck.java

```diff
@@ -0,0 +1,113 @@
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.python.checks;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+import javax.annotation.Nullable;
+
+import org.sonar.check.Rule;
+import org.sonar.plugins.python.api.PythonSubscriptionCheck;
+import org.sonar.plugins.python.api.PythonVersionUtils;
+import org.sonar.plugins.python.api.SubscriptionContext;
+import org.sonar.plugins.python.api.tree.Expression;
+import org.sonar.plugins.python.api.tree.FormatSpecifier;
+import org.sonar.plugins.python.api.tree.FormattedExpression;
+import org.sonar.plugins.python.api.tree.StringElement;
+import org.sonar.plugins.python.api.tree.StringLiteral;
+import org.sonar.plugins.python.api.tree.Tree;
+import org.sonar.python.tree.TreeUtils;
+
+@Rule(key = "S6799")
+public class FStringNestingLevelCheck extends PythonSubscriptionCheck {
+
+  private static final String MESSAGE = "Do not nest f-strings too deeply.";
+
+  private static final Set<StringElement> visited = new HashSet<>();
+
+  private static final int MAX_DEPTH = 3;
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.FILE_INPUT, ctx -> visited.clear());
+    context.registerSyntaxNodeConsumer(Tree.Kind.STRING_ELEMENT, FStringNestingLevelCheck::checkNestingDepthOfFString);
+  }
+
+  private static void checkNestingDepthOfFString(SubscriptionContext ctx) {
+    if (!supportsTypeParameterSyntax(ctx)) {
+      return;
+    }
+    StringElement element = (StringElement) ctx.syntaxNode();
+    if (isFStringNestedTooDeep(element, 0)) {
+      ctx.addIssue(element, MESSAGE);
+    }
+  }
+
+  private static boolean isFStringNestedTooDeep(StringElement element, final int count) {
+    if (!visited.contains(element) && element.isInterpolated()) {
+      visited.add(element);
+      int updatedCount = count + 1;
+      if (updatedCount >= MAX_DEPTH) {
+        return true;
+      }
+      return areFormattedExpressionsNestedTooDeep(element.formattedExpressions(), updatedCount);
+    }
+    return false;
+  }
+
+  private static boolean areFormattedExpressionsNestedTooDeep(List<FormattedExpression> formattedExpressions, int updatedCount) {
+    for (FormattedExpression formattedExpression : formattedExpressions) {
+      if (isTheNestingTooDeepInExpression(formattedExpression.expression(), updatedCount) ||
+        isTheNestingTooDeepInFormatSpecifier(formattedExpression.formatSpecifier(), updatedCount)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private static boolean isTheNestingTooDeepInExpression(Expression expression, int updatedCount) {
+    return Optional.of(expression)
+      .flatMap(TreeUtils.toOptionalInstanceOfMapper(StringLiteral.class))
+      .map(StringLiteral::stringElements)
+      .map(Collection::stream)
```

Comment:
maybe a more flat chain is better:
```java
...
.flatMap(Collections::stream)
.anyMatch(sElement -> isFStringNestedTooDeep(sElement, updatedCount))
.orElse(false);
```

Reply:
I don't think I can do this `flatMap` on the `Optional` with the stream.